`build-image` command should accept extra options and pass them to `docker build`

When running:

```
cpl build-image -a ${{ inputs.app_name }} --commit ${{ steps.vars.outputs.sha_short }} --org ${{ inputs.org }} --cache /tmp/.docker-cache
```

We get:

```
ERROR: "cpl build_image" was called with arguments ["--cache", "/tmp/.docker-cache"]
Usage: "cpl build-image -a, --app=APP_NAME"
Error: Process completed with exit code 1.
```

We should allow the `build-image` command to accept extra options and pass them to `docker build`.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,10 @@ Changes since the last non-beta release.
 
 _Please add entries here for your pull requests that are not yet released._
 
+### Changed
+
+- `build-image` command now accepts extra options and passes them to `docker build`. [PR 126](https://github.com/shakacode/heroku-to-control-plane/pull/126) by [Rafael Gomes](https://github.com/rafaelgomesxyz).
+
 ## [1.2.0] - 2024-01-03
 
 ### Fixed

--- a/docs/commands.md
+++ b/docs/commands.md
@@ -41,6 +41,7 @@ cpl apply-template gvc postgres redis rails -a $APP_NAME
 - Automatically assigns image numbers, e.g., `app:1`, `app:2`, etc.
 - Uses `.controlplane/Dockerfile` or a different Dockerfile specified through `dockerfile` in the `.controlplane/controlplane.yml` file
 - If a commit is provided through `--commit` or `-c`, it will be set as the runtime env var `GIT_COMMIT`
+- Accepts extra options that are passed to `docker build`
 
 ```sh
 cpl build-image -a $APP_NAME

--- a/lib/command/base.rb
+++ b/lib/command/base.rb
@@ -15,6 +15,9 @@ module Command
     DEFAULT_ARGS = [].freeze
     # Options for the command (use option methods below)
     OPTIONS = [].freeze
+    # Does not throw error if `true` and extra options
+    # that are not specified in `OPTIONS` are passed to the command
+    ACCEPTS_EXTRA_OPTIONS = false
     # Displayed when running `cpl help`
     # DESCRIPTION = ""
     # Displayed when running `cpl help NAME`

--- a/lib/command/build_image.rb
+++ b/lib/command/build_image.rb
@@ -7,12 +7,14 @@ module Command
       app_option(required: true),
       commit_option
     ].freeze
+    ACCEPTS_EXTRA_OPTIONS = true
     DESCRIPTION = "Builds and pushes the image to Control Plane"
     LONG_DESCRIPTION = <<~DESC
       - Builds and pushes the image to Control Plane
       - Automatically assigns image numbers, e.g., `app:1`, `app:2`, etc.
       - Uses `.controlplane/Dockerfile` or a different Dockerfile specified through `dockerfile` in the `.controlplane/controlplane.yml` file
       - If a commit is provided through `--commit` or `-c`, it will be set as the runtime env var `GIT_COMMIT`
+      - Accepts extra options that are passed to `docker build`
     DESC
 
     def call # rubocop:disable Metrics/MethodLength
@@ -32,7 +34,9 @@ module Command
       build_args = []
       build_args.push("GIT_COMMIT=#{commit}") if commit
 
-      cp.image_build(image_url, dockerfile: dockerfile, build_args: build_args)
+      cp.image_build(image_url, dockerfile: dockerfile,
+                                docker_args: config.args,
+                                build_args: build_args)
 
       progress.puts("\nPushed image to '/org/#{config.org}/image/#{image_name}'.")
     end

--- a/lib/core/controlplane.rb
+++ b/lib/core/controlplane.rb
@@ -44,12 +44,13 @@ class Controlplane # rubocop:disable Metrics/ClassLength
     api.query_images(org: a_org, gvc: a_gvc, gvc_op_type: gvc_op)
   end
 
-  def image_build(image, dockerfile:, build_args: [], push: true)
+  def image_build(image, dockerfile:, docker_args: [], build_args: [], push: true)
     # https://docs.controlplane.com/guides/push-image#step-2
     # Might need to use `docker buildx build` if compatiblitity issues arise
     cmd = "docker build --platform=linux/amd64 -t #{image} -f #{dockerfile}"
     cmd += " --progress=plain" if ControlplaneApiDirect.trace
 
+    cmd += " #{docker_args.join(' ')}" if docker_args.any?
     build_args.each { |build_arg| cmd += " --build-arg #{build_arg}" }
     cmd += " #{config.app_dir}"
     perform!(cmd)

--- a/lib/cpl.rb
+++ b/lib/cpl.rb
@@ -142,6 +142,7 @@ module Cpl
       requires_args = command_class::REQUIRES_ARGS
       default_args = command_class::DEFAULT_ARGS
       command_options = command_class::OPTIONS + ::Command::Base.common_options
+      accepts_extra_options = command_class::ACCEPTS_EXTRA_OPTIONS
       description = command_class::DESCRIPTION
       long_description = command_class::LONG_DESCRIPTION
       examples = command_class::EXAMPLES
@@ -178,7 +179,9 @@ module Cpl
                  default_args
                end
 
-        raise_args_error.call(args, nil) if (args.empty? && requires_args) || (!args.empty? && !requires_args)
+        if (args.empty? && requires_args) || (!args.empty? && !requires_args && !accepts_extra_options)
+          raise_args_error.call(args, nil)
+        end
 
         begin
           config = Config.new(args, options, required_options)

